narciuch/assignment1/src/Anagram.java

```diff
@@ -0,0 +1,123 @@
+import java.util.Arrays;
+
+public class Anagram {
+
+  /**
+   * Used in isAnagram() and isAnagramSentence()
+   * to ignore all characters not in a-z / A-Z / space.
+   */
+  private static final String replaceRegex = "[^a-zA-z ]";
+
+  /**
+   * Determines if two strings are anagrams of each other,
+   * i.e. they are made of the same letters in a different order.
+   * By default, this method is case sensitive.
+   * Input strings are unchanged.
+   * Ignores symbols as specified in the replaceRegex
+   *
+   * @param string1 : First string
+   * @param string2 : Possible anagram of the first string
+   * @return : true if string1 and string2 are anagrams of one another
+   */
+  public boolean isAnagram(String string1, String string2) {
+
+    //Remove desired characters as specified in replaceRegex
+    String str1 = string1.replaceAll(replaceRegex, "");
+    String str2 = string2.replaceAll(replaceRegex, "");
+
+    //Quick length check. Anagrams must have the same length.
+    if (str1.length() != str2.length()) {
+      return false;
+    }
+
+    //Sort and compare character by character
+    char[] chars1 = sortCharsInString(str1);
+    char[] chars2 = sortCharsInString(str2);
+    for (int i = 0; i < chars1.length; i++) {
+      if (chars1[i] != chars2[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Sorts characters within the word.
+   *
+   * @param word: the string to sort the characters of
+   * @return sorted array of chars
+   */
+  private char[] sortCharsInString(String word) {
+    char[] chars = word.toCharArray();
+    Arrays.sort(chars);
+    return chars;
+  }
+
+  /**
+   * Version of isAnagram() that can be made case insensitive
+   */
+  public boolean isAnagram(String string1, String string2, Boolean isCaseSensitive) {
+    if (!isCaseSensitive) {
+      return isAnagram(string1.toLowerCase(), string2.toLowerCase());
+    } else {
+      return isAnagram(string1, string2);
+    }
+  }
+
+  /**
+   * Determines if one sentence is an anagram of another,
+   * i.e. each word in the resulting sentence is an anagram of one of the words
+   * in the original sentence.
+   * The order of letters within each word can be shuffled as well as the order of the words.
+   * Splits up sentences on space characters.
+   * Input strings are unchanged.
+   * Ignores characters as specified in replaceRegex.
+   *
+   * @param sentence1
+   * @param sentence2, which we want to determine whether it is an anagram of sentence1.
+   * @return true if sentence2 is a sentence anagram of sentence1.
+   */
+  public boolean isAnagramSentence(String sentence1, String sentence2) {
+
+    String[] words1 = preProcessSentence(sentence1);
+    String[] words2 = preProcessSentence(sentence2);
+
+    //Compare word by word if each word in sentence2 is an anagram of the word in sentence1
+    for (int i = 0; i < words1.length; i++) {
+      if (!isAnagram(words1[i], words2[i])) {
```

Comment:
This logic assumes that sentences are anagrams only if corresponding words are anagrams, i.e. "abc def" and "cba fed" are anagrams, but "abc def" and "def abc" are not.

Original text in the assignment says "... where each word in the resulting sentence is an anagram of *one of the words* in the original sentence".

How could the code be changed to handle original requirement?

Reply:
Not quite sure what you mean here. In isAnagramSentence(), preProcessSentence() is called initially on both sentences, which splits the sentences into words and sorts the characters within each word as well as sorting the entire word array. By sorting the entire word arrays, this ensures we are comparing the desired possible word anagrams.

I have added the example you have provided "abc def" and "def abc" to the test suite and this test passes.
I have also added a comment to preProcessSentence() to explain this, hopefully this helps.

Reply:
Makes sense. Turns out that I figured out your clever optimization only partially :) That's why comments in such places (where something clever/non-obvious is done) are so important. Thanks for clarification!